docs/src/design/programs.md

```diff
@@ -53,6 +53,14 @@ When executing a *syscall* block, the VM does the following:
 
 A *syscall* block does not have any children. Thus, it must be leaf node in the tree.
 
+
+### Dyn block
+A **dyn** block is used to describe a node whose target is specified dynamically via the stack. When the VM encounters a *dyn* block, it executes a program which hashes to the target specified by the top of the stack. Thus, it is similar to both *call* and *syscall*, but it does not imply a context switch and it has a dynamic target rather than a hardcoded target. In order to execute a *dyn* block, the VM must be aware of a program with the hash value that is specified by the top of the stack. Otherwise, the execution fails.
```

Comment:
nit: I'm not sure we should say that it is similar to `call` and `syscall` - but not a strong preference from my end.